README.md

[![Arduino CI](https://github.com/RobTillaart/MATRIX7219/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/MATRIX7219/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/MATRIX7219/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/MATRIX7219/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/MATRIX7219/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/MATRIX7219/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/MATRIX7219.svg?maxAge=3600)](https://github.com/RobTillaart/MATRIX7219/releases)


# MATRIX7219

Arduino Library for controlling one or more 8x8 LED MATRIX with a MAX7219.


## Description

This experimental library provides an easy control to 8x8 MATRIX boards.
It is pretty minimalistic for now as the user still has to do bit math 
to get the right lights on as one can only set 8 LEDS at a time.

For the future two (derived) classes are planned: 
- a class that is optimized to use for a single 8x8 matrix.
- a class that buffers the state of the LEDS, allowing more functionality.


#### Tests

It is tested with an UNO and a single 8x8 matrix and a 4x 8x8 matrix-chain.
During tests it became clear that the orientation of the matrices I used, 
seen from the IN connector was different. These insights lead to the 
implementation of the **inverse, reverse** and **swap** functions.


#### Related

- TODO


#### Tested

Tested on Arduino UNO.


## Interface

**\#include "MATRIX7219.h"**

Not all functionality works, matrix parameter et al.

- **MATRIX7219(uint8_t dataPin, uint8_t selectPin, uint8_t clockPin, uint8_t matrices)** 
Constructor, initializes IO pins and the number of 8x8 matrices on same pins.
- **void begin()** resets the internals of the connected device.
- **uint8_t  getMatrixCount()** returns number of matrices set in constructor.
Convenience function.
- **void setBrightness(uint8_t bright = 2)** set bright = 0..15 for all matrices. 
Default is a relative low intensity.
- **void displayOff()** switches on low power mode == all LEDs off.
Leaves all registers the same. 
Can be used to blink the display, get attention.
- **void displayOn()** switches display on again.  
- **void displayTest(bool on = false)** sets all LEDs on full brightness.
Used to verify all LEDS still work. 
Or to implement a flashy alarm.
Default value is false ==> normal mode.
- **void clear()** clear all matrices.
- **void setRow(uint8_t row, uint8_t value, uint8_t matrix)** set a value to a certain row of a certain matrix.


#### Invert, reverse and swap.

For adapting the layout if needed.

- **void setInvert(bool invert)** invert the output. Default = false.
- **bool getInvert()** get current setting.
- **void setReverse(bool rev)** reverse the output. Default = false.
- **bool getReverse()** get current setting.
- **void setSwap(bool swap)** swaps row n with (8-n).
- **bool getSwap()** get current setting.


## Performance 

#### 0.1.0 UNO

MATRIX7219_performance.ino with UNO 16 MHz

|  board  |  function     |  time (us)  |  notes  |
|:-------:|:-------------:|:-----------:|:--------|
|   UNO   | begin         |     796     |
|   UNO   | clear         |    1572     |
|   UNO   | count         |       4     |
|   UNO   | setRow(255)   |     204     | setReverse has minimal influence
|   UNO   | setRow(0)     |     200     |
|   UNO   | setBrightness |     208     |

#### 0.1.1 UNO

Optimized registers for UNO  (for AVR from FastShiftOut).
2.5 to 4.2 times faster. (price: a few bytes storage).


|  board  |  function     |  time (us)  |  notes  |
|:-------:|:-------------:|:-----------:|:--------|
|   UNO   | begin         |     196     |  2.5x
|   UNO   | clear         |     368     |  4.2x
|   UNO   | count         |       4     |  idem
|   UNO   | setRow(255)   |      56     |  3.5x
|   UNO   | setRow(0)     |      56     |  3.5x
|   UNO   | setBrightness |      56     |  3.5x

#### 0.1.1 ESP32

First test ESP32

|  board  |  function     |  time (us)  |  notes  |
|:-------:|:-------------:|:-----------:|:--------|
|  ESP32  | begin         |      37     |
|  ESP32  | clear         |      47     |
|  ESP32  | count         |       2     |
|  ESP32  | setRow(255)   |      12     |
|  ESP32  | setRow(0)     |       8     |
|  ESP32  | setBrightness |       8     |

## Future

#### Must

- improve documentation

#### Should

- create a derived class with a buffer => 8 bytes per Matrix.
  - goal is to have **setPixel(x,y)** and **clearPixel(x,y)**
  - design first (performance)

#### Could

- examples
  - 8x8 as debugger (dump variables)
- add unit tests
  - need simulator / mock-up

#### Wont (unless requested)

- create a derived class for a single 8x8 matrix.
  - if performance gain is enough?
- store last set brightness.
  - **uint8_t getBrightness()**
  - not needed yet



CHANGELOG.md
# Change Log MATRIX7219

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.1.1] - 2023-07-29
- update AVR performance (from FastShiftOut)
- add **writeZero()** (generic improvement)
- test performance ESP32
- add clock example
- update readme.md
- update keywords.txt


## [0.1.0] - 2023-07-28
- initial version

## [0.0.1] - eons ago
- first tests




test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2023-07-28
// PURPOSE: unit tests for the MAX7219 library
//          https://github.com/RobTillaart/MATRIX7219
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// https://github.com/Arduino-CI/arduino_ci/blob/master/cpp/unittest/Assertion.h#L33-L42
// ----------------------------
// assertEqual(expected, actual)
// assertNotEqual(expected, actual)
// assertLess(expected, actual)
// assertMore(expected, actual)
// assertLessOrEqual(expected, actual)
// assertMoreOrEqual(expected, actual)
// assertTrue(actual)
// assertFalse(actual)
// assertNull(actual)
// assertNotNull(actual)

#include <ArduinoUnitTests.h>


#include "Arduino.h"
#include "MAX7219.h"


unittest_setup()
{
  fprintf(stderr, "MATRIX7219_LIB_VERSION: %s\n", (char *) MATRIX7219_LIB_VERSION);
}


unittest_teardown()
{
}


unittest(test_constructor)
{
  MATRIX7219 mx(3, 4, 5, 1);
  assertEqual(1, mx.getMatrixCount());
}


unittest_main()


//  -- END OF FILE --



MATRIX7219.h
#pragma once
//
//    FILE: MATRIX7219.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.1
//    DATE: 2023-07-28
// PURPOSE: Arduino Library for 8x8 LED MATRIX MAX7219
//     URL: https://github.com/RobTillaart/MATRIX7219


#include "Arduino.h"


#define MATRIX7219_LIB_VERSION        (F("0.1.1"))


class MATRIX7219
{
public:
  MATRIX7219(uint8_t dataPin, uint8_t selectPin, uint8_t clockPin, uint8_t matrices = 1);

  uint8_t  getMatrixCount();
  void     begin();

  //  bright = 0..15
  void     setBrightness(uint8_t bright = 2);
  void     displayOff();
  void     displayOn();
  void     displayTest(bool on = false);

  void     clear();
  //  row   = 1..8
  //  value = 0..255
  void     setRow(uint8_t row, uint8_t value, uint8_t matrix);

  void     setInvert(bool invert);
  bool     getInvert();
  void     setReverse(bool reverse);
  bool     getReverse();
  void     setSwap(bool swap);
  bool     getSwap();

private:

  void     _write(uint8_t b);
  void     _writeZero();         //  optimized  writing of all 0
  uint8_t  _reverse8(uint8_t in);


#if defined(ARDUINO_ARCH_AVR) || defined(ARDUINO_ARCH_MEGAAVR)

  volatile uint8_t *_dataOutRegister;
  uint8_t _dataOutBit;

  volatile uint8_t *_clockRegister;
  uint8_t _clockBit;

#endif
  
  uint8_t  _dataPin;
  uint8_t  _selectPin;
  uint8_t  _clockPin;
  uint8_t  _matrices;

  bool     _invert = false;
  bool     _reverse = false;
  bool     _swap = false;
};



//  -- END OF FILE -- 



MATRIX7219.cpp
//
//    FILE: MATRIX7219.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.1
//    DATE: 2023-07-28
// PURPOSE: Arduino Library for 8x8 LED MATRIX MAX7219
//     URL: https://github.com/RobTillaart/MATRIX7219


#include "MATRIX7219.h"

#define MATRIX7219_DECODE_MODE     0x09
#define MATRIX7219_BRIGHTNESS      0x0A
#define MATRIX7219_SCAN_LIMIT      0x0B
#define MATRIX7219_SHUT_DOWN       0x0C
#define MATRIX7219_DISPLAY_TEST    0x0F


MATRIX7219::MATRIX7219(uint8_t dataPin, uint8_t selectPin, uint8_t clockPin, uint8_t matrices)
{
  _dataPin   = dataPin;
  _selectPin = selectPin;
  _clockPin  = clockPin;
  _matrices  = matrices;

   pinMode(_dataPin,  OUTPUT);
   pinMode(_selectPin,OUTPUT);
   pinMode(_clockPin, OUTPUT);

   digitalWrite(_dataPin,  HIGH);
   digitalWrite(_selectPin,HIGH);
   digitalWrite(_clockPin, HIGH);

#if defined(ARDUINO_ARCH_AVR) || defined(ARDUINO_ARCH_MEGAAVR)
  uint8_t _port    = digitalPinToPort(_dataPin);
  _dataOutRegister = portOutputRegister(_port);
  _dataOutBit      = digitalPinToBitMask(_dataPin);

  _port            = digitalPinToPort(_clockPin);
  _clockRegister   = portOutputRegister(_port);
  _clockBit        = digitalPinToBitMask(_clockPin);
#endif
}


uint8_t MATRIX7219::getMatrixCount()
{
  return _matrices;
}


void MATRIX7219::begin()
{
  for (int m = 0; m < _matrices; m++)
  {
    digitalWrite(_selectPin, LOW);
    _write(MATRIX7219_SCAN_LIMIT);
    _write(0x07);                       //  use bits 0..7
    digitalWrite(_selectPin, HIGH);
  }
  for (int m = 0; m < _matrices; m++)
  {
    digitalWrite(_selectPin, LOW);
    _write(MATRIX7219_DECODE_MODE);
    _writeZero();                       //  No decode for digits 7–0
    digitalWrite(_selectPin, HIGH);
  }
  for (int m = 0; m < _matrices; m++)
  {
    digitalWrite(_selectPin, LOW);
    _write(MATRIX7219_SHUT_DOWN);
    _write(0x01);                       //  display on
    digitalWrite(_selectPin, HIGH);
  }
  for (int m = 0; m < _matrices; m++)
  {
    digitalWrite(_selectPin, LOW);
    _write(MATRIX7219_DISPLAY_TEST);
    _writeZero();                       //  normal mode
    digitalWrite(_selectPin, HIGH);
  }
}


void MATRIX7219::setBrightness(uint8_t bright)
{
  if (bright > 15) bright = 15;

  digitalWrite(_selectPin, LOW);
  for (int m = 0; m < _matrices; m++)
  {
    _write(MATRIX7219_BRIGHTNESS);
    _write(bright);
  }
  digitalWrite(_selectPin, HIGH);
}


void MATRIX7219::displayOff()
{
  digitalWrite(_selectPin, LOW);
  for (int m = 0; m < _matrices; m++)
  {
    _write(MATRIX7219_SHUT_DOWN);
    _writeZero();
  }
  digitalWrite(_selectPin, HIGH);
}


void MATRIX7219::displayOn()
{
  digitalWrite(_selectPin, LOW);
  for (int m = 0; m < _matrices; m++)
  {
    _write(MATRIX7219_SHUT_DOWN);
    _write(0x01);
  }
  digitalWrite(_selectPin, HIGH);
}


void MATRIX7219::displayTest(bool on)
{
  digitalWrite(_selectPin, LOW);
  for (int m = 0; m < _matrices; m++)
  {
    _write(MATRIX7219_DISPLAY_TEST);
    if (on) _write(0x01);
    else _writeZero();
  }
  digitalWrite(_selectPin, HIGH);
}


void MATRIX7219::clear()
{
  for (uint8_t row = 1; row < 9; row++)
  {
    digitalWrite(_selectPin, LOW);
    for (int m = 0; m < _matrices; m++)
    {
      _write(row);
      _writeZero();
    }
    digitalWrite(_selectPin, HIGH);
  }
}


void MATRIX7219::setRow(uint8_t row, uint8_t value, uint8_t matrix)
{
  digitalWrite(_selectPin, LOW);
  for (int m = _matrices; m > matrix; m--)
  {
    _writeZero();
    _writeZero();
  }

  if (_swap) row = 9 - row;
  _write(row);
  if (_invert)  value = 255 - value;
  if (_reverse) value = _reverse8(value);
  _write(value);

  for (int m = matrix - 1; m > 0; m--)
  {
    _writeZero();
    _writeZero();
  }
  digitalWrite(_selectPin, HIGH);
}


///////////////////////////////////////////////////////
//
//  LAYOUT
//
void MATRIX7219::setInvert(bool invert)
{
  _invert = invert;
}


bool MATRIX7219::getInvert()
{
  return _invert;
}


void MATRIX7219::setReverse(bool rev)
{
  _reverse = rev;
}


bool MATRIX7219::getReverse()
{
  return _reverse;
}


void MATRIX7219::setSwap(bool swap)
{
  _swap = swap;
}


bool MATRIX7219::getSwap()
{
  return _swap;
}


///////////////////////////////////////////////////////
//
//  PRIVATE
//
void MATRIX7219::_write(uint8_t b)
{
#if defined(ARDUINO_ARCH_AVR) || defined(ARDUINO_ARCH_MEGAAVR)

  uint8_t cbmask1  = _clockBit;
  uint8_t cbmask2  = ~_clockBit;
  uint8_t outmask1 = _dataOutBit;
  uint8_t outmask2 = ~_dataOutBit;

  for (uint8_t mask = 0x80; mask > 0; mask >>= 1)
  {
    uint8_t oldSREG = SREG;
    noInterrupts();
    if ((b & mask) == 0) *_dataOutRegister &= outmask2;
    else                 *_dataOutRegister |= outmask1;
    *_clockRegister |= cbmask1;
    *_clockRegister &= cbmask2;
    SREG = oldSREG;
  }

#else
  
  uint8_t clk = _clockPin;
  uint8_t dat = _dataPin;
  for (uint8_t mask = 0x80; mask > 0; mask >>= 1)
  {
    digitalWrite(clk, LOW);
    digitalWrite(dat, (b & mask) > 0);
    digitalWrite(clk, HIGH);
  }

#endif
}


//  optimization
void MATRIX7219::_writeZero()
{
#if defined(ARDUINO_ARCH_AVR) || defined(ARDUINO_ARCH_MEGAAVR)

  uint8_t cbmask1  = _clockBit;
  uint8_t cbmask2  = ~_clockBit;
  uint8_t outmask1 = _dataOutBit;
  uint8_t outmask2 = ~_dataOutBit;
  
  *_dataOutRegister &= outmask2;
  
  for (uint8_t mask = 0x80; mask > 0; mask >>= 1)
  {
    uint8_t oldSREG = SREG;
    noInterrupts();
    *_clockRegister |= cbmask1;
    *_clockRegister &= cbmask2;
    SREG = oldSREG;
  }

#else
  
  uint8_t clk = _clockPin;
  digitalWrite(_dataPin, LOW);
  for (uint8_t mask = 0x80; mask > 0; mask >>= 1)
  {
    digitalWrite(clk, LOW);
    digitalWrite(clk, HIGH);
  }
  
#endif
}


//  From CRC 0.3.3
uint8_t MATRIX7219::_reverse8(uint8_t in)
{
  uint8_t x = in;
  x = (((x & 0xAA) >> 1) | ((x & 0x55) << 1));
  x = (((x & 0xCC) >> 2) | ((x & 0x33) << 2));
  x =          ((x >> 4) | (x << 4));
  return x;
}

//  -- END OF FILE --

